src/java.base/share/classes/jdk/internal/constant/ConstantUtils.java

```diff
@@ -269,59 +271,46 @@ public static String dropFirstAndLastChar(String s) {
         return s.substring(1, s.length() - 1);
     }
 
-    /**
-     * Parses a method descriptor string, and return a list of field descriptor
-     * strings, return type first, then parameter types
-     *
-     * @param descriptor the descriptor string
-     * @return the list of types
-     * @throws IllegalArgumentException if the descriptor string is not valid
-     */
-    public static List<ClassDesc> parseMethodDescriptor(String descriptor) {
-        int cur = 0, end = descriptor.length();
-        ArrayList<ClassDesc> ptypes = new ArrayList<>();
-        ptypes.add(null); // placeholder for return type
-
-        if (cur >= end || descriptor.charAt(cur) != '(')
-            throw new IllegalArgumentException("Bad method descriptor: " + descriptor);
-
-        ++cur;  // skip '('
-        while (cur < end && descriptor.charAt(cur) != ')') {
-            int len = skipOverFieldSignature(descriptor, cur, end, false);
-            if (len == 0)
-                throw new IllegalArgumentException("Bad method descriptor: " + descriptor);
-            ptypes.add(resolveClassDesc(descriptor, cur, len));
-            cur += len;
-        }
-        if (cur >= end)
-            throw new IllegalArgumentException("Bad method descriptor: " + descriptor);
-        ++cur;  // skip ')'
-
-        int rLen = skipOverFieldSignature(descriptor, cur, end, true);
-        if (rLen == 0 || cur + rLen != end)
-            throw new IllegalArgumentException("Bad method descriptor: " + descriptor);
-        ptypes.set(0, resolveClassDesc(descriptor, cur, rLen));
-        return ptypes;
+    public static PrimitiveClassDescImpl forPrimitiveType(String descriptor, int offset) {
+        return switch (descriptor.charAt(offset)) {
+            case JVM_SIGNATURE_BYTE    -> CD_byte;
+            case JVM_SIGNATURE_CHAR    -> CD_char;
+            case JVM_SIGNATURE_FLOAT   -> CD_float;
+            case JVM_SIGNATURE_DOUBLE  -> CD_double;
+            case JVM_SIGNATURE_INT     -> CD_int;
+            case JVM_SIGNATURE_LONG    -> CD_long;
+            case JVM_SIGNATURE_SHORT   -> CD_short;
+            case JVM_SIGNATURE_VOID    -> CD_void;
+            case JVM_SIGNATURE_BOOLEAN -> CD_boolean;
+            default -> throw badMethodDescriptor(descriptor);
+        };
     }
 
-    private static ClassDesc resolveClassDesc(String descriptor, int start, int len) {
+    static ClassDesc resolveClassDesc(String descriptor, int start, int len) {
         if (len == 1) {
-            return Wrapper.forPrimitiveType(descriptor.charAt(start)).basicClassDescriptor();
+            return forPrimitiveType(descriptor, start);
         }
-        // Pre-verified in parseMethodDescriptor; avoid redundant verification
+
+        // objectDesc appears a lot during the bootstrap process, so optimize it
+        String objectDesc = "Ljava/lang/Object;";
+        if (len == objectDesc.length() && descriptor.regionMatches(start, objectDesc, 0, len)) {
```

Comment:
Note that from my bytestack investigations, `regionMatches` can be CPU-intensive like hashCode calculation. Running this trick against `Ljava/lang/String;` (same length) might cause a lot of misses and waste a lot of CPU time. Can you try on a case with many `Ljava/lang/String;` descriptors and see the results compared to the build without this special case?

Reply:
Also `Double` `Module` `Number` `Record` `Thread` are susceptible to taking more time in `regionMatches`.

Reply:
I think you can have the current patch RFR (ready for review) without this deduplication given my concerns. We can add deduplication of frequent descriptors in another patch after confirming it is a net positive.

Reply:
regionMatches uses vectorizedMismatch for comparison. The length of `Ljava/lang/Object;` is 18. In the code without intrinsic, it will exit after two long comparisons. If there is vector128 optimization, it will exit after one comparison, and the overhead will be very small.

Reply:
I think you can add a case where the parameters are all these vulnerable names like `Ljava/lang/Objecq;` and repeat 8 times, and see how big a negative impact these close mismatches have.

Reply:
The scenario of `Ljava/lang/Objecq` does not exist. I have added the following:
```
(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;
(Ljava/lang/Integer;Ljava/lang/Integer;)Ljava/lang/Integer;
```

In these two scenarios, the length of `Ljava/lang/String;` is the same as that of `Ljava/lang/Object;`, and the length of `Ljava/lang/Integer;` is different from that of `Ljava/lang/Object;`, but the performance is similar. This proves that the performance overhead of regionMatches is very small.

Reply:
Here are the performance numbers running under Aliyun ECS c8i (my MacBook M1 Pro is not available at this time, so the performance test is run on Alibaba Cloud’s ECS)

* CPU Intel®Xeon®Emerald Rapids (Linux x64)

## Benchmark Script

```sh
# current
git checkout ee321904e6eee307f27cb38a70ac4ca75a8f446b
make test TEST="micro:java.lang.constant.MethodTypeDescFactories.ofDescriptor"
```

## Benchmark Numbers
```
Benchmark                                                                            (descString)  Mode  Cnt     Score    Error  Units
MethodTypeDescFactories.ofDescriptor                      (Ljava/lang/Object;Ljava/lang/String;)I  avgt    6    73.913 ±  0.708  ns/op
MethodTypeDescFactories.ofDescriptor                                                          ()V  avgt    6     1.729 ±  0.013  ns/op
MethodTypeDescFactories.ofDescriptor     (Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;  avgt    6    85.204 ±  0.337  ns/op
MethodTypeDescFactories.ofDescriptor     (Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;  avgt    6   120.307 ±  1.887  ns/op
MethodTypeDescFactories.ofDescriptor  (Ljava/lang/Integer;Ljava/lang/Integer;)Ljava/lang/Integer;  avgt    6   120.832 ±  1.904  ns/op
MethodTypeDescFactories.ofDescriptor                                         ()Ljava/lang/Object;  avgt    6    23.709 ±  0.548  ns/op
MethodTypeDescFactories.ofDescriptor                     ([IJLjava/lang/String;Z)Ljava/util/List;  avgt    6   109.046 ±  1.712  ns/op
MethodTypeDescFactories.ofDescriptor                                        ()[Ljava/lang/String;  avgt    6    30.997 ±  0.556  ns/op
MethodTypeDescFactories.ofDescriptor                                                     (..IIJ)V  avgt    6   179.165 ±  2.371  ns/op
MethodTypeDescFactories.ofDescriptor                                            ([III.Z[B..[.[B).  avgt    6   529.011 ± 12.875  ns/op
MethodTypeDescFactories.ofDescriptor                                     (.....................).  avgt    6  1690.000 ±  9.597  ns/op
```

This performance number proves my point above that regionMatches has low overhead.